File: CSharp/Core/Common.WinForms/Helpers/DispatcherHelper.cs
namespace System.Windows.Threading;

public static class DispatcherHelper
{
    public static void Execute(Action action)
    {
        // no cross-thread concerns
        if (Application.OpenForms.Count == 0)
        {
            action.Invoke();
			return;
        }

		try
		{
			if (Application.OpenForms[0]!.InvokeRequired)
                // Marshall to Main Thread
				Application.OpenForms[0]?.Invoke(action);
            else
                // We are already on the Main Thread
                action.Invoke();
        }
		catch (Exception)
		{
			// ignore as might be thrown on shutting down
		}
    }
}

File: VB/Core/CommonVB.WinForms/Helpers/DispatcherHelper.vb
Public Module DispatcherHelper

    Public Sub Execute(action As Action)

        ' no cross-tread concerns
        If Application.OpenForms.Count = 0 Then
            action.Invoke()
            Return
        End If

        Try

            If Application.OpenForms(0).InvokeRequired Then
                ' Marshall to Main Thread
                Application.OpenForms(0).Invoke(action)
            Else
                ' no cross-tread concerns
                action.Invoke()
            End If

        Catch ex As Exception

            ' ignore as might be thrown on shutting down

        End Try

    End Sub

End Module